client/src/escrow_client/mod.rs

```diff
@@ -1,155 +1,152 @@
+use std::sync::Arc;
+
+use crate::common::model::EscrowRegistration;
+use cdk::nuts::Token;
+
 use super::*;
-use cashu_escrow_common::nostr::PubkeyMessage;
-use cdk::nuts::PublicKey;
 
-pub enum Trader {
-    Buyer(EscrowUser),
-    Seller(EscrowUser),
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum TradeMode {
+    Buyer,
+    Seller,
 }
 
-pub struct EscrowUser {
-    pub escrow_coordinator_npub: String,
-    pub escrow_pk_ts: (PublicKey, Timestamp),
-    pub contract: TradeContract,
-    pub wallet: EcashWallet,
-    pub nostr_client: NostrClient,
+pub struct InitEscrowClient {
+    ecash_wallet: ClientEcashWallet,
+    escrow_contract: TradeContract,
+    trade_mode: TradeMode,
 }
 
-impl Trader {
-    pub async fn init_trade(&self) -> anyhow::Result<()> {
-        match self {
-            Trader::Buyer(config) => {
-                self.buyer_pipeline(config).await?;
-                Ok(())
-            }
-            Trader::Seller(config) => {
-                self.seller_pipeline(config).await?;
-                Ok(())
-            }
+/// Initial Escrow Client state.
+impl InitEscrowClient {
+    pub fn new(
+        ecash_wallet: ClientEcashWallet,
+        escrow_contract: TradeContract,
+        trade_mode: TradeMode,
+    ) -> Self {
+        Self {
+            ecash_wallet,
+            escrow_contract,
+            trade_mode,
         }
     }
 
-    async fn buyer_pipeline(&self, config: &EscrowUser) -> anyhow::Result<()> {
-        let token = config.wallet.create_escrow_token(config).await?;
-        dbg!("Sending token to the seller: {}", token.as_str());
-
-        config
-            .nostr_client
-            .submit_trade_token_to_seller(&config.contract.npub_seller, &token)
+    /// The trade initialization is the same for both buyer and seller.
+    ///
+    /// After this the coordinator data is set, state trade registered.
+    ///
+    /// After this state the trade contract is effectfull as well, possible coordinator fees must be payed.
+    pub async fn register_trade(
+        &self,
+        nostr_client: Arc<NostrClient>,
+    ) -> anyhow::Result<RegisteredEscrowClient> {
+        let nostr_client_ref = nostr_client.clone();
+        let reg_msg_fut =
+            tokio::spawn(async move { nostr_client_ref.receive_escrow_message(20).await });
+
+        let coordinator_pk = &self.escrow_contract.npubkey_coordinator;
+        let contract_message = serde_json::to_string(&self.escrow_contract)?;
+        dbg!("sending contract to coordinator...");
+        nostr_client
+            .client
+            .send_private_msg(*coordinator_pk, &contract_message, None)
             .await?;
 
-        // either send signature or begin dispute
-        Ok(())
+        let registration_message = reg_msg_fut.await??;
+        let escrow_registration = serde_json::from_str(&registration_message)?;
```

Comment:
I think we have to verify the EscrowRegistration ID (contract hash) here on client side. Otherwise i imagine a malicious client could submit two, malicious contracts to the coordinator and if the client doesn't verify that the contract id matches, would opt in to the wrong contract?

Reply:
Addressed separately in Issue https://github.com/f321x/cashu-escrow-kit/issues/10